### ISSUE TITLE
OOP refactoring - loader.php

### ISSUE THREAD

**Issue by redaxmedia** [redaxmedia/redaxscript#46]
This issue affects includes/loader.php...

Let us learn from Zend Framework, we should split it into classes like:

- [x] Assetic/Loader (equals the current <code>loader</code> function)
- [x] Assetic/Transport (equals the current <code>scripts_transport</code> function)
- [x] Extend Head bundle with like useCache() methods that set a flag to use the Loader
- [x] <del>Assetic/Minifier (class does already exists in root)</del>
- [x] Use CSSNano to combine CSS to xxx.min.css packages
 - [x] base
 - [x] templates
- [x] Use UglifyJS to combine JS to xxx.min.js packages
 - [x] base
 - [x] templates

Edit: There is no elegant way to minify assets with PHP, this is basicly the job of build tools like CSSNano and UglifyJS running with NodeJS - therefore we remove the Minifier, the Loader can do basic concatenation of files, this is enough for us

This refactoring is related to issue #166 because collecting of files should be the job of someone else living in it's own namespace. Therefore the Asset functionality will be used by the <code>Head/Script</code> and <code>Head/Style</code> class.

## Actual state:

1. There are [.loader](https://github.com/redaxmedia/redaxscript/blob/master/templates/default/scripts/.loader) files in each template

<pre>
[inherit]
#
# templates can inherit from the .loader file other templates, in this case from admin
#
1 = admin

[single]
#
# this files will not be combined and rendered via script src="..." tag
#
1 = //ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js
2 = scripts/fallback.js

[template_init]
#
# template assets that will be combined and rendered inside the the head tag
# https://github.com/redaxmedia/redaxscript/blob/master/templates/default/head.phtml#L24
#
1 = scripts/support.js
2 = scripts/init.js
3 = templates/default/scripts/init.js

[template]
#
# template assets that will be combined and rendered before body tag closing
# https://github.com/redaxmedia/redaxscript/blob/master/templates/default/index.phtml#L49
#
1 = scripts/dialog.js
2 = scripts/form.js
3 = scripts/interface.js
4 = scripts/misc.js

[settings]
#
# render inline via script tag or via ?p=loader route
# rewrite image urls
# minify output
#
deploy = inline
rewrite = false
minify = true
</pre>
1. The [loader](https://github.com/redaxmedia/redaxscript/blob/master/includes/loader.php#L18) function is collecting all the .loader files and merges files as needed
1. Depending on the "deploy" setting it renders a script tag with inline code or a src="loader/scripts"
1. The route "loader/scripts" and "loader/styles" are not cached and generated on each request - [index.php](https://github.com/redaxmedia/redaxscript/blob/3.0.0/index.php#L28-L31) and [install.php](https://github.com/redaxmedia/redaxscript/blob/3.0.0/install.php#L16-L19) still have the code for a global routing
1. The template tags for rendering [scripts](https://github.com/redaxmedia/redaxscript/blob/master/includes/loader.php#L292) and [styles](https://github.com/redaxmedia/redaxscript/blob/master/includes/loader.php#L212) tags are inside the loader.php
1. The [transport](https://github.com/redaxmedia/redaxscript/blob/master/includes/loader.php#L393) of PHP variables to JS variables are inside the loader.php

## Target state:

1. First of all, forget the old code and idea of .loader files - do not copy and paste code!!!
1. Collecting (equals the .loader config) of CSS and JS files is no longer a task of the loader
1. Rendering script and style tags is no longer task of the loader
1. The Assetic/Loader should be able to retrieve a config of array and does the job of the current <code>loader</code> function what basicly means combining files via include and returning the result
1. The Assetic/Loader should take care of caching by dropping files to <code>/cache/{hash}</code> - the hash is being created from the config array keys and values - so it should be unique
1. Assetic/Transport can transport PHP variables to JS variables via inline Javascript - keep all the DEFAULT transports for selected registry, whole language, baseURL, generator and version
1. <del>The Assetic/Minifier is for minifying scripts and styles</del>
1. The bundle of Assetic will be consumed by the <code>Head/Style</code> and <code>Head/Script</code> classes, those classes provide the config array of files to be included, merged, minified and cached

**What happens if the Assetic/Loader cannot create the cache files?**
 1. The whole thing is going to break?
 1. There is a route like /loader/{hash} fallback to fetch the result like in the actual state?
 1. The Head/Script and Head/Link is using a inline script and style tags?
 1. The Head/Loader nothing and therefore output each file in a single tag?

Suggestion how to start:

+ Use the TDD approach
+ Do not start the Assetic bundle before the Head bundle
+ Write the Assetic classes and methods according to needs of the Head bundle, if you implement Template::scripts()->minify() you have to write the Assetic/Minifier class ... and so on
+ Write unit tests for the Assetic even if code parts are covered by the Head bundle
+ Insure that the default transported variables are present: https://snag.gy/CNTlbm.jpg

Todo for integration:

- [x] Remove all the .loader files inside the templates
- [x] Remove the global routing in index.php and install.php
- [x] Remove Minifier class and related testing stuff

**Comment by redaxmedia** [comment]
Done
